core/matrix/csr.cpp

```diff
@@ -83,23 +84,13 @@ void Csr<ValueType, IndexType>::apply_impl(const LinOp *alpha, const LinOp *b,
 
 
 template <typename ValueType, typename IndexType>
-std::unique_ptr<Coo<ValueType, IndexType>> Csr<ValueType, IndexType>::make_coo()
-    const
+void Csr<ValueType, IndexType>::convert_to(
+    Coo<ValueType, IndexType> *result) const
 {
     auto exec = this->get_executor();
     auto tmp = Coo<ValueType, IndexType>::create(
         exec, this->get_size(), this->get_num_stored_elements());
-    exec->run(csr::make_convert_row_ptrs_to_idxs(
-        this->get_const_row_ptrs(), this->get_size()[0], tmp->get_row_idxs()));
-    return tmp;
-}
-
-
-template <typename ValueType, typename IndexType>
-void Csr<ValueType, IndexType>::convert_to(
-    Coo<ValueType, IndexType> *result) const
-{
-    auto tmp = this->make_coo();
+    exec->run(csr::make_convert_to_coo(tmp.get(), this));
     tmp->values_ = this->values_;
     tmp->col_idxs_ = this->col_idxs_;
```

Comment:
Can you move the values before calling the `conver_to_coo` kernel? Just in case someone wants to access the sizes of the Arrays?

Reply:
I did that for both directions. Note that I also had to change the coo_kernels.cpp /.cu slightly since the source matrix is no longer complete when calling the convert_to_csr kernel.